core/Symbols.cc

```diff
@@ -1171,6 +1171,57 @@ vector<Symbol::RequiredAncestor> Symbol::requiredAncestors(const GlobalState &gs
     return res;
 }
 
+// All required ancestors by this class or module
+vector<Symbol::RequiredAncestor> requiredAncestorsTransitiveInternal(const GlobalState &gs, SymbolRef sym,
+                                                                     std::vector<SymbolRef> seen) {
+    std::vector<Symbol::RequiredAncestor> res;
+
+    if (std::find(seen.begin(), seen.end(), sym) != seen.end()) {
+        return res; // Break recursive loops if we already visited this ancestor
+    }
+    seen.emplace_back(sym);
+
+    for (auto ancst : sym.data(gs)->requiredAncestors(gs)) {
+        ancst.origin = sym;
+        res.emplace_back(ancst);
+        for (auto sancst : requiredAncestorsTransitiveInternal(gs, ancst.symbol, seen)) {
+            if (sancst.symbol == sym) {
+                continue;
+            }
+            res.emplace_back(sancst);
+        }
+    }
+
+    auto parent = sym.data(gs)->superClass();
+    if (parent.exists()) {
+        for (auto ancst : requiredAncestorsTransitiveInternal(gs, parent, seen)) {
+            if (ancst.symbol == sym) {
+                continue;
+            }
+            res.emplace_back(ancst);
+        }
+    }
+
+    for (auto mixin : sym.data(gs)->mixins()) {
+        for (auto ancst : mixin.data(gs)->requiredAncestors(gs)) {
+            if (ancst.symbol == sym) {
+                continue;
+            }
+            ancst.origin = mixin;
+            res.emplace_back(ancst);
+        }
+    }
+
+    return res;
+}
+
+// All required ancestors by this class or module
+vector<Symbol::RequiredAncestor> Symbol::requiredAncestorsTransitive(const GlobalState &gs) const {
+    ENFORCE(this->isClassOrModule(), "Symbol is not a class or module: {}", this->show(gs));
+    std::vector<SymbolRef> seen;
+    return requiredAncestorsTransitiveInternal(gs, this->ref(gs), seen);
```

Comment:
With a little bit of refactoring, isn't `seen` basically what you're trying to return? You could thread it through as a reference, and use `absl::find_if` with a lambda to project the values out of `seen`.

Reply:
Not sure how you would do it.

Be the following example:

```rb
module RA; end

module M
  requires_ancestor RA
end

class C
  include M
end
```

`seen` would contain `C, M, RA` (ie all the thing we *traversed*) while `res` would contain only `RA` (ie all the thing we *collected*).

Am I missing something?

Reply:
I see. Well, my comment still stands about copying with `seen`. You could thread references to both the seen vector, and the `res` vector through, and then you wouldn't end up re-allocating/copying quite so much.

Reply:
I've already done it for `seen`, but yeah, good idea about `res`, I'll change this.